### PR TITLE
Phase 6: Ensemble Weight Optimization — Learned Model Weights

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Our current 8-seed ensemble uses equal weights (1/8 each). Some models may be systematically better on certain splits. Optimizing per-model weights (constrained to be non-negative and sum to 1) could improve ensemble performance at ZERO training cost.

Additionally, test selective ensembles: maybe the best 5-of-8 models beat all 8 (removing high-variance outlier models).

## Instructions

**Step 1 — Get ensemble predictions:**
The 8 baseline model W&B run IDs are: rboyvjeo, h0uog211, kwt8tw52, 5j26p5v1, rmump7ke, ujt9cu0l, 7fw8ksxq, 0lsry8km

Use eval_ensemble.py (or modify it) to:
1. Load each model's checkpoint from W&B
2. Run inference on ALL validation splits (in_dist, ood_cond, tandem_transfer, ood_re)
3. Save per-sample predictions for each model

**Step 2 — Optimize weights:**
Write a script that:
1. Loads per-sample predictions from all 8 models
2. Uses scipy.optimize.minimize (or simple grid search) to find weights w_1..w_8 that minimize the surface pressure MAE on a held-out portion of in_dist validation
3. Evaluate the weighted ensemble on all splits
4. Compare: equal-weight vs optimized-weight vs best-N-of-8

**Step 3 — Also test N-of-8 subsets:**
Evaluate all C(8,N) combinations for N=4,5,6,7 and find the best subset for each split.

**GPU allocation:**
This is a post-hoc analysis experiment. You should need at most 1-2 GPUs for inference. Use the remaining GPUs for additional seed training:

| GPU 0-1 | Ensemble weight optimization (inference only) |
| GPU 2-7 | Train 6 new standard 3L seeds (90-95) for future ensemble expansion |

Seed training command (same as baseline):
```bash
for seed in 90 91 92 93 94 95; do
  CUDA_VISIBLE_DEVICES=$((seed - 88)) python train.py --agent askeladd \
    --wandb_group phase6/ensemble-weight-opt \
    --wandb_name "askeladd/3L-s${seed}" \
    --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
    --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
    --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
    --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
    --seed ${seed} &
done
wait
```

**What to report:**
- Equal-weight ensemble MAE vs optimized-weight ensemble MAE
- Best N-of-8 subsets and their MAE
- The learned weight vector

## Baseline
Current 8-seed ensemble (equal weights, PR #2076): p_in=12.4, p_oodc=6.7, p_tan=29.4, p_re=5.8
W&B: wandb-applied-ai-team/senpai-v1, run IDs: rboyvjeo, h0uog211, kwt8tw52, 5j26p5v1, rmump7ke, ujt9cu0l, 7fw8ksxq, 0lsry8km

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Status Update — Checkpoint Availability

The 8 original baseline model checkpoints (run IDs: rboyvjeo, h0uog211, kwt8tw52, 5j26p5v1, rmump7ke, ujt9cu0l, 7fw8ksxq, 0lsry8km) are not available locally — they were trained in a previous session and models are saved locally, not as W&B artifacts.

**Workaround:** I have 4 baseline models from the srf4L experiment (seeds 42-45, run IDs: hfyp045n, cwnsbzq3, j6nx4oje, ll7rta5t) and 6 new seeds (90-95) currently training. I'll run the weight optimization analysis on these 10 models (4+6) once the new seeds complete (~10 min remaining).

This still fully tests the hypothesis (optimized weights vs equal weights, best-N subsets) — just with different individual models. The optimization methodology is identical.

Seeds 90-95 training on GPUs 2-7, weight optimization will run on GPU 0 after seed training completes.

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Setup

The original 8 baseline checkpoints were not available locally (trained in a prior session, not saved as W&B artifacts). Instead, I trained 10 fresh models with identical asinh baseline flags:
- Seeds 42-45 (from srf4L experiment, baseline arm)
- Seeds 90-95 (new, trained for this PR)

All 10 models use `surface_refine_layers=3`, `cosine_T_max=160`, `asinh_pressure`, `asinh_scale=0.75`, and all other baseline flags.

### Individual Model Performance (p_in)

| Seed | Run ID | p_in |
|------|--------|------|
| 42 | hfyp045n | 13.6 |
| 43 | cwnsbzq3 | 13.0 |
| 44 | j6nx4oje | 13.7 |
| 45 | ll7rta5t | 12.9 |
| 90 | ici6bxi1 | 13.0 |
| 91 | 6chuzqal | 12.8 |
| 92 | xcsqiwdv | 13.0 |
| 93 | sxisuynb | 12.9 |
| 94 | q8m1w63d | 13.8 |
| 95 | ggn5mioe | 13.4 |

### Ensemble Weight Optimization Results

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| **Equal-weight (10/10)** | **12.27** | **6.67** | **29.08** | **5.83** |
| **Optimised weights (SLSQP)** | **12.27** | **6.67** | **29.08** | **5.83** |
| Best 4-of-10 | 12.08 | 6.89 | 29.40 | 5.85 |
| Best 5-of-10 | 12.08 | 6.81 | 29.31 | 5.84 |
| Best 6-of-10 | 12.07 | 6.76 | 29.18 | 5.80 |
| Best 7-of-10 | 12.10 | 6.72 | 29.14 | 5.81 |

### Optimised Weight Vector

```
[0.1004, 0.0994, 0.0993, 0.0998, 0.1020, 0.1010, 0.0993, 0.0988, 0.1011, 0.0990]
```

The optimiser converged to essentially equal weights (all within ±1% of 0.1). No model is systematically better enough to warrant higher weight.

### Best N-of-10 Subsets (selected by in_dist surf_p MAE)

| N | Indices | Run IDs | Seeds |
|---|---------|---------|-------|
| 4 | [3,4,5,7] | ll7rta5t, ici6bxi1, 6chuzqal, sxisuynb | 45, 90, 91, 93 |
| 5 | [1,3,4,5,7] | + cwnsbzq3 | + 43 |
| 6 | [1,3,4,5,6,7] | + xcsqiwdv | + 92 |
| 7 | [1,3,4,5,6,7,9] | + ggn5mioe | + 95 |

Consistently excluded: seeds 42 (p_in=13.6), 44 (p_in=13.7), 94 (p_in=13.8) — the weakest individual models.

### Comparison vs PR Baseline (8-seed ensemble, PR #2076)

| Metric | PR #2076 (8-seed) | This (10-seed equal) | Best 6-of-10 |
|--------|-------------------|---------------------|--------------|
| p_in | 12.4 | **12.27** (-1.0%) | **12.07** (-2.7%) |
| p_oodc | 6.7 | **6.67** (-0.4%) | 6.76 (+0.9%) |
| p_tan | 29.4 | **29.08** (-1.1%) | 29.18 (-0.7%) |
| p_re | 5.8 | **5.83** (+0.5%) | **5.80** (0%) |

### Commands Used

```bash
# Seed training (GPUs 2-7)
for seed in 90 91 92 93 94 95; do
  CUDA_VISIBLE_DEVICES=$((seed - 88)) python train.py --agent askeladd \
    --wandb_group phase6/ensemble-weight-opt \
    --wandb_name "askeladd/3L-s${seed}" \
    --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
    --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
    --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
    --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
    --seed ${seed} &
done

# Weight optimization (GPU 0)
CUDA_VISIBLE_DEVICES=0 python weight_opt.py \
  --run_ids hfyp045n cwnsbzq3 j6nx4oje ll7rta5t ici6bxi1 6chuzqal xcsqiwdv sxisuynb q8m1w63d ggn5mioe \
  --asinh_scale 0.75 --wandb_name "askeladd/ensemble-weight-opt-10models"
```

W&B: run `99gz442v` (weight opt analysis), group `phase6/ensemble-weight-opt` (seed training)

### What Happened

**Weight optimization provides essentially zero benefit.** The SLSQP optimizer converged to near-equal weights (~0.1 each), indicating no model is systematically better across samples. This makes sense: all models share the same architecture and training procedure, differing only in random seed. Each model's errors are essentially random around the same mean, so no model deserves more weight.

**Selective ensembles (best-N-of-K) show a small benefit.** Removing the 3 weakest models (seeds 42, 44, 94 — the ones with highest p_in) gives a marginal improvement. Best 6-of-10 achieves p_in=12.07 vs 12.27 for equal-weight-all — a ~1.6% improvement from model selection alone.

**More models helps most.** The 10-model ensemble slightly beats the original 8-model ensemble (p_in 12.27 vs 12.4, p_tan 29.08 vs 29.4). Adding more diverse seeds provides diminishing but real returns.

### Suggested Follow-ups

- **Model checkpoints should be saved as W&B artifacts** so future analysis can access any prior model. Currently only saved locally, making cross-session analysis impossible.
- **Selective ensemble at inference time**: A simple heuristic — drop any model with single-seed p_in > 13.5 — effectively implements the "best N-of-K" without exhaustive search.
- **More seeds** would continue to reduce variance. Each additional seed provides ~sqrt(1/N) improvement on ensemble MAE.
- Since weight optimisation provides no benefit, future ensemble work should focus on model diversity (different architectures, loss functions) rather than weight tuning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2089 because ADVISOR: Excellent analysis. Key findings:

1. **Weight optimization: NULL RESULT.** SLSQP converged to near-equal weights (~0.1 each). This confirms that same-architecture models trained with different seeds don't benefit from non-uniform weighting — errors are random around the same mean.

2. **10-model ensemble (12.27 p_in) is flat vs current baseline (12.2 p_in, seeds 66-73).** Marginal.

3. **Best-6-of-10 selective ensemble** (p_in=12.07) shows promise but trades off p_oodc (+0.9%). Not a clean win.

4. **Most valuable insight:** 'Future ensemble work should focus on model diversity (different architectures, loss functions) rather than weight tuning.' Fully agree — this motivates our new research direction.

Seeds 90-95 run IDs: ici6bxi1, 6chuzqal, xcsqiwdv, sxisuynb, q8m1w63d, ggn5mioe — available for future ensemble expansion.

Closing and assigning you a new experiment..
